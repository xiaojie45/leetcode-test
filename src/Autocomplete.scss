.Autocomplete {
  width: 300px;
  margin: 0 auto;
  padding: 42px 24px 0 50px;
  .search-input {
    color: rgba(0,0,0,0.65);
    border-radius: 4px;
    position: relative;
    input {
      width: 300px;
      line-height: 30px;
      box-sizing: border-box;
    }
    span {
      display: inline-block;
      position: absolute;
      right: 10px;
      top: 50%;
      transform: translateY(-50%);
    }
    img {
      height: 24px;
      vertical-align: middle;
    }
  }
  .search-results {
    height: 200px;
    overflow: auto;
    border-radius: 4px;
    border: 1px solid #d9d9d9;
    &.hide {
      display: none
    }
    div {
      height: 40px;
      line-height: 40px;
      &:hover {
        background-color: #ccc
      }
    }
  }
  .search-error {
    color: red
  }
}
