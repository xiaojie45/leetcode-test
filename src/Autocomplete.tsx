import React from 'react'
import { Subject, from } from 'rxjs'
import { switchMap, debounceTime, filter, takeUntil } from 'rxjs/operators'

import { MAXSEARCHLENGTH, URL } from './constant'
import spinnerIcon from './spinner.gif'
import searchIcon from './search.svg'
import './Autocomplete.scss'

const result: string[] = []
const initialState = {
	search: '',
	result,
	error: '',
	loading: false,
	hideResult: true
}
type State = Readonly<typeof initialState>

class Autocomplete extends React.Component<any, State> {
	state: State = initialState
	private search$: Subject<string> = new Subject()

	componentDidMount() {
		this.search$ = new Subject()
		this.getAutoSearch()
	}

	getSearchList = (word: string) => {
		this.setState({ loading: true })
		return fetch(`${URL}&search=${word}`).then((res) => res.json())
	}

	getAutoSearch = () => {
		/*istanbul ignore next*/
		const observer = {
			next: (value: Array<Array<string>>) => {
				this.setState({
					result: value[1],
					hideResult: false,
					loading: false
				})
			},
			complete: () => {
				this.setState({
					error: '您输入的字符数过多',
					loading: false
				})
			}
		}
		const exceedMaxLength$ = this.search$.pipe(
			filter((val: string) => val.length > MAXSEARCHLENGTH)
		)
		/*istanbul ignore next*/
		this.search$
			.pipe(
				debounceTime(500),
				switchMap((e: string) => from(this.getSearchList(e))),
				takeUntil(exceedMaxLength$)
			)
			.subscribe(observer)
	}

	onBlur = (e: React.ChangeEvent<HTMLInputElement>) => {
		this.setState({ hideResult: true })
	}

	onSearch = (e: React.ChangeEvent<HTMLInputElement>) => {
		const search = e.target.value
		this.setState({ search })
		search && this.search$.next(search)
		if (search.length < MAXSEARCHLENGTH && this.state.error) {
			this.setState({ error: '' })
			this.getAutoSearch()
		}
	}

	render() {
		const { search, error, loading, hideResult, result } = this.state
		const searchBar = loading ? spinnerIcon : searchIcon
		return (
			<div className="Autocomplete">
				<div className="search-input">
					<input
						type="text"
						onBlur={this.onBlur}
						onChange={this.onSearch}
						value={search}
					/>
					<span>
						<img src={searchBar} alt="search icon" />
					</span>
				</div>
				{error ? (
					<div className="search-error">{error}</div>
				) : (
					<div className={`search-results ${hideResult && 'hide'}`}>
						{result.map((r: string) => {
							return <div key={r}>{r}</div>
						})}
					</div>
				)}
			</div>
		)
	}
}

export default Autocomplete
