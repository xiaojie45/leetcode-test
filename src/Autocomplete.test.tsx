import React from 'react'
import Enzyme from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import { Subject } from 'rxjs'

import Autocomplete from './Autocomplete'

Enzyme.configure({ adapter: new Adapter() })
const { shallow } = Enzyme
const result = [
	'上海市',
	'上海地铁',
	'上海浦东国际机场',
	'上海交通大学',
	'上海虹桥国际机场',
	'上海话',
	'上海迪士尼樂園',
	'上海轨道交通2号线',
	'上海“11·15”特别重大火灾',
	'上海轨道交通1号线'
]

describe('autocomplete component', () => {
	const normalEventObj = {
		target: {
			value: 'shanghai'
		}
	}
	let search$: Subject<string>
	let wrapper: any

	beforeEach(() => {
		wrapper = shallow(<Autocomplete />)
		search$ = new Subject()
		search$.subscribe({
			next: (value) => {
				wrapper.setState({
					result: value || [],
					hideResult: false,
					loading: false
				})
				expect(wrapper.state(['result'])).toBe(value)
				expect(wrapper.state(['hideResult'])).toBe(false)
				expect(wrapper.state(['loading'])).toBe(false)
			},
			complete: () => {
				wrapper.setState({
					error: '您输入的字符数过多',
					loading: false
				})
				expect(wrapper.state(['error'])).toBe('您输入的字符数过多')
				expect(wrapper.state(['loading'])).toBe(false)
			}
		})
	})

	it('autocomplete should render correctly', () => {
		expect(wrapper).toMatchSnapshot()
	})

	it('input change should change state search', () => {
		wrapper.find('input').simulate('change', normalEventObj)
		expect(wrapper.state(['search'])).toBe('shanghai')
		expect(wrapper.state(['error'])).toBe('')
	})

	it('should hide when input field blur', () => {
		wrapper.find('input').simulate('blur')
		expect(wrapper.state(['hideResult'])).toBe(true)
	})

	it('should have at most results shown', () => {
		wrapper.setState({ result })
		expect(wrapper.find('.search-results div').length).toBe(10)
	})

	it('should have at most results shown', () => {
		wrapper.setState({ error: '您输入的字符数过多' })
		wrapper.find('input').simulate('change', normalEventObj)
		expect(wrapper.state(['error'])).toBe('')
	})

	it('should have loading icon spinning when calls api', () => {
		const instance = wrapper.instance()
		instance.getSearchList('上海')
		expect(wrapper.state(['loading'])).toBe(true)
	})

	it('should show result when it gets value', () => {
		search$.next(normalEventObj.target.value)
	})

	it('should show error when it completes', () => {
		search$.complete()
	})
})
