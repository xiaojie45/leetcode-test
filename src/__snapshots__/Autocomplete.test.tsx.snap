// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`autocomplete component autocomplete should render correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Autocomplete />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="search-input"
        >
          <input
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value=""
          />
          <span>
            <img
              alt="search icon"
              src="search.svg"
            />
          </span>
        </div>,
        <div
          className="search-results hide"
        />,
      ],
      "className": "Autocomplete",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <input
              onBlur={[Function]}
              onChange={[Function]}
              type="text"
              value=""
            />,
            <span>
              <img
                alt="search icon"
                src="search.svg"
              />
            </span>,
          ],
          "className": "search-input",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "onBlur": [Function],
              "onChange": [Function],
              "type": "text",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <img
                alt="search icon"
                src="search.svg"
              />,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "alt": "search icon",
                "src": "search.svg",
              },
              "ref": null,
              "rendered": null,
              "type": "img",
            },
            "type": "span",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [],
          "className": "search-results hide",
        },
        "ref": null,
        "rendered": Array [],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="search-input"
          >
            <input
              onBlur={[Function]}
              onChange={[Function]}
              type="text"
              value=""
            />
            <span>
              <img
                alt="search icon"
                src="search.svg"
              />
            </span>
          </div>,
          <div
            className="search-results hide"
          />,
        ],
        "className": "Autocomplete",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <input
                onBlur={[Function]}
                onChange={[Function]}
                type="text"
                value=""
              />,
              <span>
                <img
                  alt="search icon"
                  src="search.svg"
                />
              </span>,
            ],
            "className": "search-input",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "onBlur": [Function],
                "onChange": [Function],
                "type": "text",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <img
                  alt="search icon"
                  src="search.svg"
                />,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "alt": "search icon",
                  "src": "search.svg",
                },
                "ref": null,
                "rendered": null,
                "type": "img",
              },
              "type": "span",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [],
            "className": "search-results hide",
          },
          "ref": null,
          "rendered": Array [],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
